HSEARCH-3490 Add experimental support for Elasticsearch 7.0.0.beta1

yrodiere opened the pull request:
https://hibernate.atlassian.net//browse/HSEARCH-3490

I ran a full build (ES5, ES6, ES7) there: http://ci.hibernate.org/blue/organizations/jenkins/hibernate-search-yoann/detail/HSEARCH-3490/4/

yrodiere commented:
Force-pushed an update to address some Sonar warnings.

Note the loss in coverage is a bit artificial: coverage numbers only takes into account the "default" build, i.e. the one with ES6, not the one with ES7.

gsmet commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/dialect/impl/es6/Elasticsearch6Dialect.java:
```diff
@@ -7,28 +7,38 @@
-		 */
-		return new Elasticsearch6WorkBuilderFactory( gsonProvider );
+	public ElasticsearchSearchResultExtractorFactory createSearchResultExtractorFactory() {
+		// Necessary because the total hit count it formatted differently in ES7
```
s/it formatted/is formatted/

gsmet commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/dialect/impl/es7/Elasticsearch7Dialect.java:
```diff
@@ -0,0 +1,46 @@
+import com.google.gson.GsonBuilder;
+
+/**
+ * The dialect for Elasticsearch 6.
```
7

yrodiere commented:
I addressed your comments, rebased and force-pushed. Waiting for CI because of the rebase.

> A bit worried by the amount of work required to support a new version of ES. But I suppose that's how it is, considering how they deal with backward compatibility.

On the bright side, the actual changes are only `813 insertions(+), 456 deletions(-)`, the rest being tests and test resources (init scripts, ...), and even for tests there's the introduction of test dialects which will hopefully represent less work next time, now that it's there.

yrodiere commented:
Merged, thanks!
